src/core2/graph.js

```diff
@@ -313,8 +313,16 @@ export class Graph {
     plugins: Plugins,
     graphs: $ReadOnlyArray<Graph>
   ): Graph {
-    const _ = {plugins, graphs};
```

Comment:
Hmm, I notice that we could instead have an API that doesn't require passing plugins, but instead takes the set union of plugins in the merged graphs. I posit that would basically always be "the right thing", the current api basically forces the user to separately track which plugins are in the merged graph, with the penalty that the call will fail if any necessary plugins were excluded. What do you think of making that change?

Reply:
It’s not clear to me that there is a meaningful definition of set union
for plugin lists, because it’s not clear that there’s a meaningful
definition for equality of plugin handlers. Suppose that `h1` and `h2`
are two plugin handlers with the same `.pluginName()`. It is clear that
in some cases we must throw a “duplicate plugin handler” error, and in
others we must recognize that `h1` and `h2` are “the same” and pick one
of them arbitrarily. But how do we make this choice?

We can’t rely on reference equality: if we have
```js
const g1 = new Graph([new GitPluginHandler()]);
const g2 = new Graph([new GitPluginHandler(), new GithubPluginHandler()]);
```
then `g1.plugins()[0] !== g2.plugins()[0]`.

We can’t rely on class/`.constructor` equality, for two reasons. First,
it would be totally reasonable to define
```js
class FooPluginHandler {
  constructor(someImportantOptions) {
    // …
  }
  // …
}
```
such that two instances of `FooPluginHandler` could really be
meaningfully different. Second, it is also totally reasonable to define
```js
export function barPluginHandler() {
  return {
    createRef: /* … */,
    createPayload: /* … */,
  };
}
```
such that a plugin handler is just an `Object`.

One attempt at a resolution would be to soft-mandate that plugins expose
not a class implementing `PluginHandler` but instead singleton instance
of such a class, with the understanding that this exact instance should
be used to represent the plugin.

What do you think?

Reply:
I like the "singleton instance" proposal. We can then assume (and document) that all plugins with the same name are assumed equivalent, and if multiple distinct instances are in fact provided, we choose between them arbitrarily.

Reply:
Thanks for coming up with a thoughtful and interesting example of when we'd want to pass options from the plugin object. When we do add options like you described, I think we'd need to serialize them into the graph, otherwise graph semantics are not portable / reproducible. So we'd probably come up with some way of storing pluginOptions in the graph, and graph equality (and thus merge conservative's conservativism) would explicitly respect those pluginOptions (which would, of course, also have compat handlers). Thus, I think that even in light of your points, mergeConservative should eventually not require a plugins argument.

However, you've persuaded me that this merits careful attention, and there's no need to figure this all out now. Given that we have been quite good about refactoring APIs, I'm fine approving this with its current API and we can take another pass on it as we develop the plugin serialization story.